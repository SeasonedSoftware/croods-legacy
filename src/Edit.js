import { Component } from 'react'
import { connect } from 'react-redux'

import withOptions from './withOptions'
import mapStateToProps from './mapStateToProps'
import mapDispatchToProps from './mapDispatchToProps'
import setOrFetchInfo from './setOrFetchInfo'

class Edit extends Component {
  constructor(props) {
    super(props)
    setOrFetchInfo(props)
  }

  componentWillReceiveProps(nextProps) {
    const { id } = nextProps
    const { id: oldId } = this.props

    if (id.toString() !== oldId.toString()) {
      setOrFetchInfo(nextProps)
    }
  }

  componentDidUpdate(prevProps) {
    const { actions, updated } = this.props
    const { updated: oldUpdated } = prevProps

    if (updated && !oldUpdated) {
      actions.resetUpdated()
    }
  }

  render() {
    const { render, renderUpdated, actions, info, fetchingInfo } = this.props
    const { infoError, updated, updating, updateError: error } = this.props
    const { renderLoading, renderError } = this.props
    const { update } = actions

    if (infoError) {
      return renderError(infoError)
    }

    if (!info || fetchingInfo) {
      return renderLoading()
    }

    if (updated) {
      return renderUpdated(updated)
    }

    return render({ info, update, updating, error })
  }
}

export default withOptions(
  connect(
    mapStateToProps,
    mapDispatchToProps,
  )(Edit),
)
