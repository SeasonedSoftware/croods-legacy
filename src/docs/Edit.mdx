---
name: Edit
route: /edit
order: 6

---

# Edit

The Edit component is a union of Info and New.
This component does a `GET` request to the `path` like Info and its update function,
when executed, does a `PUT` request with the jsonified parameter as request body, like New.

## API

| Property            | Type          | Required | Default          | Description                                                                                                                          |
| ------------------- | ------------- | -------- | ---------------- | ------------------------------------------------------------------------------------------------------------------------------------ |
| id                  | String/Number | true     | -                | Ex: 1234 or '1234'                                                                                                                   |
| name                | String        | true     | -                | It must be equal to the reducer name set in your `rootReducer.js` and will be used as `path` if none is provided Ex: `baseUrl/name`. |
| render              | Func          | true     | -                | A function returning a React Node. Ex: ({ info = {}, update = (id, ...attributes), updating = bool, error }, props) -> < JSX />      |
| renderUpdated       | Func          | false    | -                | A function returning a React Node. Ex: (updated = {}) -> < JSX />                                                                    |
| parentId            | String/Number | false    | -                | Parent element's Id, for more specificity in changes.                                                                                |
| path                | String        | false    | `/${name}/${id}` | Ex: /foo/123                                                                                                                         |
| parseInfoResponse   | Func          | false    | -                | Parse specific responses to adjust the API to croods patterns. Ex: (json, response, requestAttributes) -> Object                     |
| parseUpdateResponse | Func          | false    | -                |                                                                                                                                      |

## How to use

Don't forget to add a reducer with the same name used here in your rootReducer, with **createReducer**

```javascript
[...]
import { Edit } from 'croods'

export default (
  <Edit
    name="pets"
    id="123"
    // if path is different than /pets/${id}, you can set a custom path like so:
    // path='/customPath'
    render={({ info, update, updating, error }) => (
      <Form onSubmit={update} submitting={updating} initialValues={info} />
    )}
  />
)

```

### render

The render function receives four parameters: **info, update, updating and error**.

- info: the response body of the `GET` request.
- update: the function to make the `PUT` request.
- updating: a boolean that identifies when the `PUT` request is being done and it is not finished.
- error: The error message in case that update fails.

### renderUpdated

In order to do something after update, use renderUpdated function to
redirect to other page or show the resource updated/some message/anything in the current page.
This function receives the response body as parameter.

## Example

You can see a working example using this component in [here](https://github.com/SeasonedSoftware/croods/blob/master/example/src/colors/Edit.js).
