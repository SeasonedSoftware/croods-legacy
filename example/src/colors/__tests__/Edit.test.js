import React from 'react'
import renderer from 'react-test-renderer'

import Edit from '../Edit'

jest.mock('croods', () => ({
  Edit: props => (
    <div {...props}>
      Edit -{' '}
      <div>
        render -{' '}
        {props.render({
          info: Object,
          update: Function,
          updating: Boolean,
          error: String,
        })}
      </div>
      <div>renderUpdated - {props.renderUpdated({ id: 1 })}</div>
    </div>
  ),
}))

jest.mock('../Form', () => props => (
  <div {...props}>Form - {props.children}</div>
))
jest.mock('react-router-dom/Link', () => props => (
  <div {...props}>Link - {props.children}</div>
))
jest.mock('react-router-dom/Redirect', () => props => (
  <div {...props}>Redirect - {props.children}</div>
))

it('renders correctly', () => {
  const props = {
    match: {
      params: {
        id: 1,
      },
    },
  }
  const tree = renderer.create(<Edit {...props} />).toJSON()
  expect(tree).toMatchSnapshot()
})
