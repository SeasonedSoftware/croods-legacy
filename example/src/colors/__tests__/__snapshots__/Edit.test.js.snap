// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  id={1}
  name="colors"
  render={[Function]}
  renderUpdated={[Function]}
>
  Edit -
   
  <div>
    render -
     
    <h1>
      Object
    </h1>
    <div
      initialValues={[Function]}
      onSubmit={[Function]}
      submitting={[Function]}
    >
      Form - 
    </div>
    <div
      to="/"
    >
      Link - 
      Back
    </div>
  </div>
  <div>
    renderUpdated - 
    <div
      to="/1"
    >
      Redirect - 
    </div>
  </div>
</div>
`;
